snuba/subscriptions/data.py

```diff
@@ -18,8 +19,17 @@ class InvalidSubscriptionError(Exception):
     pass
 
 
+PartitionId = NewType("PartitionId", int)
+
+
 @dataclass(frozen=True)
-class Subscription:
+class SubscriptionIdentifier:
+    partition: PartitionId
+    uuid: UUID
+
+
+@dataclass(frozen=True)
+class SubscriptionData:
```

Comment:
I think we can come up with a better name than this, but I haven't figured out one that make sense yet.

Reply:
Something query related makes sense to me, although it does also include the resolution, which is how frequently we actually run the query, and that seems less about the query itself and more about the subscription.

Reply:
(tl;dr: I think it is hard to come up with a good name for this because it has to work within the context of/on top of some abstractions that are starting to crack a bit. Half-baked thoughts and non-answer below.)

My only hesitation with a name like that is that it seems to suggest that this data structure is somehow substitutable with a `Query`. Not only is it not substitutable with `Query`, this is closer related a `Request`, since it uses `from_date` and `to_date` to generate the time range and enforce the `project_id` requirement. These requirements are probably fairly easily to relax, and will probably _have_ to be relaxed for the data science use cases that we've briefly discussed. I don't think we necessarily need to solve this right now, but we should probably start figuring out how this is going to evolve in the near-future. I do think that dealing with `Query` at this level will be generally more flexible than `Request`.

Another way of thinking about subscriptions that I have been kicking around a bit is that with subscriptions, we're basically introducing unbound variables into the query — specifically `timestamp` and `offset` (and we could also introduce `partition` if that's useful. These variables are then bound by the executor when the query is run based on external state (the tick data.) If we were able to include these variables in the query itself (assuming the new AST form) as `Variable` nodes that get bound and transformed to `Literal` types at some stage of processing/execution, that seems like the most flexible (and powerful) way to think about these queries.

Reply:
`I do think that dealing with Query at this level will be generally more flexible than Request.` Query in the sense of `a concept of Query` yes, totally. I would advise against using the current Query object that is mutable and gets mutated in the process of running the query.

Reply:
`f we were able to include these variables in the query itself (assuming the new AST form) as Variable` technically a piece of cake.
Have you ever considered a possible use case for unbound variable in standard point in time queries ? Cannot think about a case. But yes, binding variable during query execution is definitively feasible

Reply:
> Have you ever considered a possible use case for unbound variable in standard point in time queries ? Cannot think about a case.

The only use cases I could think of would be:

1. prepared statements: not sure there is much justification for this, but if we start to do more advanced query planning it might be helpful, I also would expect there to be pushback on this idea because it's not as flexible for the end-user
2. ability to parameterize queries: this doesn't absolve us of the need to have good injection protections so it's probably not worth the effort to expose an interface for binding variables

Reply:
Going to leave this as-is for now for the sake of expediency (and because I don't think anything more specific is a great fit right now), but I agree it should be changed at some point — probably when addressing some of the broader use cases like mentioned in https://github.com/getsentry/snuba/pull/706#discussion_r370863578.